### ISSUE TITLE
MethodGraph incorrectly prefers abstract method from base class to default method from interface

### ISSUE THREAD

**Issue by robbytx** [raphw/byte-buddy#949]
Given a class hierarchy where some method's implementation exists strictly as a default method in a Java 8 interface:
```java
public interface Interface {

    String getType();
}

public static abstract class BaseClass implements Interface {

}

public interface SubInterfaceWithType extends Interface {

    default String getType() {
        return "SubInterfaceWithType";
    }
}

public static class ConcreteClass extends BaseClass implements SubInterfaceWithType {

}
```

ByteBuddy incorrectly passes the abstract method rather than the default method when delegating to the implementation.  This issue appears to stem from the fact that the MethodGraph compiler incorrectly the abstract method from the superclass over the default method implementation inherited from the interface.

Using the class hierarchy above, and the test code below, I get the following result (where this issue is denoted by ``INCORRECTLY``):
```
For class example.Test$ConcreteClass:
  public abstract java.lang.String example.Test$Interface.getType()
returns:
  SubInterfaceWithType

For class example.Test$ConcreteClass$ByteBuddy$2K32eamD:
  public java.lang.String example.Test$ConcreteClass$ByteBuddy$2K32eamD.getType()
returns:
  INCORRECTLY intercepted ABSTRACT method: public abstract java.lang.String example.Test$Interface.getType()

----------------------------------------

For interface example.Test$Interface:
  net.bytebuddy.dynamic.scaffold.MethodGraph$Linked$Delegation@8a6fb820
correctly contains:
  public abstract java.lang.String example.Test$Interface.getType()

For class example.Test$BaseClass:
  net.bytebuddy.dynamic.scaffold.MethodGraph$Linked$Delegation@863996eb
correctly contains:
  public abstract java.lang.String example.Test$Interface.getType()

For interface example.Test$SubInterfaceWithType:
  net.bytebuddy.dynamic.scaffold.MethodGraph$Linked$Delegation@bfa0171d
correctly contains:
  public java.lang.String example.Test$SubInterfaceWithType.getType()

For class example.Test$ConcreteClass:
  net.bytebuddy.dynamic.scaffold.MethodGraph$Linked$Delegation@28abaec0
INCORRECTLY contains:
  public abstract java.lang.String example.Test$Interface.getType()
```

Here the full example code:
```java
package example;

import net.bytebuddy.ByteBuddy;
import net.bytebuddy.ClassFileVersion;
import net.bytebuddy.description.method.MethodDescription;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.dynamic.scaffold.MethodGraph;
import net.bytebuddy.implementation.bind.annotation.AllArguments;
import net.bytebuddy.implementation.bind.annotation.BindingPriority;
import net.bytebuddy.implementation.bind.annotation.Origin;
import net.bytebuddy.implementation.bind.annotation.RuntimeType;
import net.bytebuddy.implementation.bind.annotation.StubValue;
import net.bytebuddy.implementation.bind.annotation.SuperCall;
import net.bytebuddy.implementation.bind.annotation.This;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.concurrent.Callable;

import static net.bytebuddy.implementation.MethodDelegation.to;

public class Test {

    public interface Interface {

        String getType();
    }

    public static abstract class BaseClass implements Interface {

    }

    public interface SubInterfaceWithType extends Interface {

        default String getType() {
            return "SubInterfaceWithType";
        }
    }

    public static class ConcreteClass extends BaseClass implements SubInterfaceWithType {

    }

    /**
     * Mimicks Mockito's own <a href="https://github.com/mockito/mockito/blob/893e2f476445ce273d82ec73a93ced713610df86/src/main/java/org/mockito/internal/creation/bytebuddy/MockMethodInterceptor.java#L126">DispatcherDefaultingToRealMethod</a>
     */
    public static class DispatcherDefaultingToRealMethod {

        @SuppressWarnings("unused")
        @RuntimeType
        @BindingPriority(BindingPriority.DEFAULT * 2)
        public static Object interceptSuperCallable(
                @This Object mock,
                @Origin Method invokedMethod,
                @AllArguments Object[] arguments,
                @SuperCall(serializableProxy = true) Callable<?> superCall)
                throws Throwable {
            return "correctly intercepted actual method: " + invokedMethod;
        }

        @SuppressWarnings("unused")
        @RuntimeType
        public static Object interceptAbstract(
                @This Object mock,
                @StubValue Object stubValue,
                @Origin Method invokedMethod,
                @AllArguments Object[] arguments)
                throws Throwable {
            return "INCORRECTLY intercepted ABSTRACT method: " + invokedMethod;
        }
    }

    public static void main(String[] args) throws IllegalAccessException, InstantiationException, NoSuchMethodException {
        Class<? extends ConcreteClass> original = ConcreteClass.class;
        Class<? extends ConcreteClass> intercepted = new ByteBuddy(ClassFileVersion.JAVA_V8)
                .subclass(original)
                .method(any -> true)
                .intercept(to(DispatcherDefaultingToRealMethod.class))
                .make()
                .load(original.getClassLoader())
                .getLoaded();

        for (Class<? extends ConcreteClass> type : Arrays.asList(original, intercepted)) {
            ConcreteClass instance = type.newInstance();
            System.out.println("For " + type + ":");
            System.out.println("  " + type.getMethod("getType"));
            System.out.println("returns:");
            System.out.println("  " + instance.getType());
            System.out.println();
        }

        System.out.println("----------------------------------------\n");

        for (Class<?> type : Arrays.asList(Interface.class, BaseClass.class, SubInterfaceWithType.class, ConcreteClass.class)) {
            MethodGraph.Linked graph = MethodGraph.Compiler.DEFAULT.compile(TypeDescription.ForLoadedType.of(type));
            MethodDescription methodDescription = graph.listNodes().asMethodList().stream()
                    .filter(candidate -> candidate.getName().equals("getType"))
                    .findFirst().orElseThrow(IllegalStateException::new);
            System.out.println("For " + type + ":");
            System.out.println("  " + graph);
            if (SubInterfaceWithType.class.isAssignableFrom(type) && methodDescription.isAbstract()) {
                System.out.print("INCORRECTLY ");
            } else {
                System.out.print("correctly ");
            }
            System.out.println("contains:");
            System.out.println("  " + methodDescription);
            System.out.println();
        }

        System.out.println();
    }

}
```

Among other potential impact, this issue appears to causes [Mockito spies](https://github.com/mockito/mockito/wiki/Using-Spies-(and-Fakes)) to return `null` from such methods rather invoking the actual implementation.

**Comment by robbytx** [comment]
By the way, I just want to say that the ByteBuddy library is extremely impressive -- in its capabilities, its relative ease of use, and its impact in the JVM ecosystem -- so thank you for the work you have put into building this!

Please consider this issue simply as a "heads up" -- I have no expectation that this will be fixed on any particular timeline, but I simply wanted to make you and others aware of this problem, and I wanted to document the issue for myself related to my own workaround.

**Comment by raphw** [comment]
Thanks, appreciate it!

This is a bug indeed. I'll look into it once I find some time. This opens another point where ambiguous resolutions can become non-ambiguous again if an interface subclasses two interfaces that both implement a method with a similar signature.

I reporduced it in a test case and will certainly fix it some time!

**Comment by robbytx** [comment]
FWIW - I've also become aware of https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8062389, which (as I understand) causes `Class.getMethod` to return the incorrect method on Java 8 JREs.  The only reason I mention here is to make sure you avoid using the result of `Class.getMethod` on Java 8 as the definitive answer for what MethodGraph should return.

**Comment by robbytx** [comment]
I've found what I think is another instance of this issue, but that does not involve default methods at all.  

At least, it seems like ByteBuddy is ignoring methods in subinterfaces that match (by argument type) methods implemented in the superclass, but that do NOT match in terms of return type.

Here's a test case that demonstrates the problem:
```java
package example;

import net.bytebuddy.ByteBuddy;
import net.bytebuddy.ClassFileVersion;
import net.bytebuddy.description.method.MethodDescription;
import net.bytebuddy.implementation.FixedValue;
import net.bytebuddy.matcher.ElementMatcher;

import java.lang.reflect.Method;

import static net.bytebuddy.matcher.ElementMatchers.isDeclaredBy;
import static net.bytebuddy.matcher.ElementMatchers.named;

public class Test2 {

    public interface ReturnType {
    }

    public interface ExtendedReturnType extends ReturnType {
    }

    public interface Interface {

        ReturnType getValue();
    }

    public static class BaseClass implements Interface {

        @Override
        public ReturnType getValue() {
            return new ReturnType() {
                @Override
                public String toString() {
                    return "ReturnType from BaseClass";
                }
            };
        }
    }

    public interface SubInterfaceWithExtendedReturnType extends Interface {

        ExtendedReturnType getValue();
    }

    public static void main(String[] args) throws Exception {
        Interface original = new BaseClass();
        System.out.println(original.getClass() + " returns: " + original.getValue());
        System.out.println();

        System.out.println("When matching only methods in the SubInterfaceWithExtendedReturnType interface:");
        generateSubclass(isDeclaredBy(SubInterfaceWithExtendedReturnType.class));
        System.out.println();

        System.out.println("When matching any method named getValue:");
        generateSubclass(named("getValue"));
        System.out.println();
    }

    private static void generateSubclass(ElementMatcher.Junction<MethodDescription> methodMatcher) throws InstantiationException, IllegalAccessException, NoSuchMethodException {
        Class<? extends BaseClass> subclass = new ByteBuddy(ClassFileVersion.JAVA_V8)
                .subclass(BaseClass.class)
                .implement(SubInterfaceWithExtendedReturnType.class)
                .method(methodMatcher)
                .intercept(FixedValue.value(new ExtendedReturnType() {
                    @Override
                    public String toString() {
                        return "ExtendedReturnType from intercepted type";
                    }
                }))
                .make()
                .load(BaseClass.class.getClassLoader())
                .getLoaded();
        SubInterfaceWithExtendedReturnType instance = (SubInterfaceWithExtendedReturnType) subclass.newInstance();
        try {
            System.out.println(instance.getClass() + " returns: " + instance.getValue());
        } catch (Throwable t) {
            System.err.println(instance.getClass() + " throws:");
            t.printStackTrace();
        }

        System.out.println(subclass + "'s getValue method is " + subclass.getMethod("getValue"));

        Method[] declaredMethods = subclass.getDeclaredMethods();
        System.out.println(subclass + " declares following " + declaredMethods.length + " methods:");
        for (Method declaredMethod : declaredMethods) {
            System.out.println(declaredMethod + " with modifiers: 0x" + Integer.toHexString(declaredMethod.getModifiers()));
        }
    }
}
```

When I run this on Java 8, I see the following output:
```
class example.Test2$BaseClass returns: ReturnType from BaseClass

When matching only methods in the SubInterfaceWithExtendedReturnType interface:
class example.Test2$BaseClass$ByteBuddy$Lq7skHq4 throws:
java.lang.AbstractMethodError: Method example/Test2$BaseClass$ByteBuddy$Lq7skHq4.getValue()Lexample/Test2$ExtendedReturnType; is abstract
	at example.Test2$BaseClass$ByteBuddy$Lq7skHq4.getValue(Unknown Source)
	at example.Test2.generateSubclass(Test2.java:81)
	at example.Test2.main(Test2.java:52)
class example.Test2$BaseClass$ByteBuddy$Lq7skHq4's getValue method is public example.Test2$ReturnType example.Test2$BaseClass.getValue()
class example.Test2$BaseClass$ByteBuddy$Lq7skHq4 declares following 0 methods:

When matching any method named getValue:
class example.Test2$BaseClass$ByteBuddy$vbd7wf7d returns: ExtendedReturnType from intercepted type
class example.Test2$BaseClass$ByteBuddy$vbd7wf7d's getValue method is public example.Test2$ExtendedReturnType example.Test2$BaseClass$ByteBuddy$vbd7wf7d.getValue()
class example.Test2$BaseClass$ByteBuddy$vbd7wf7d declares following 2 methods:
public example.Test2$ExtendedReturnType example.Test2$BaseClass$ByteBuddy$vbd7wf7d.getValue() with modifiers: 0x1041
public example.Test2$ReturnType example.Test2$BaseClass$ByteBuddy$vbd7wf7d.getValue() with modifiers: 0x1

```

In the first call to `generateSubclass(isDeclaredBy(SubInterfaceWithExtendedReturnType.class))`, observe that:
* NO methods are generated on the subclass, even though `SubInterfaceWithExtendedReturnType` clearly declares the `ExtendedReturnType getValue()` method, which is NOT declared/implemented by the superclass.
* The generated class is loaded and instantiated fine, but:
* We get an `AbstractMethodError` when trying to invoke the `getValue` method, because:
* The only available `getValue` method is the one provided by the base class, which has an incompatible return type.

In the second call to `generateSubclass(named("getValue")), observe that:
* The method is implemented this time (because the matcher returns true for `BaseClass.getValue()`), but:
* The declared method that actually holds the implementation is the one that overrides `BaseClass.getValue()`, while:
* The declared method that returns `ExtendedReturnType` is a synthetic+bridge method (modifiers contains 0x1040).

To summarize:
1. My expectation is that ByteBuddy's `DynamicType.Builder.method(ElementMatcher<MethodDescription>)` would be invoked at some point with the `SubInterfaceWithExtendedReturnType.getValue()` method (that returns `ExtendedReturnType`, and this is not occurring in `1.10.17`.
2. The manifestation of this problem appears very similar to the default method problem I've reported in this issue, and therefore I'm assuming it has the same root cause.  If this is not the case, please let me know, and I'm happy to report this as a separate top-level issue.

**Comment by raphw** [comment]
This is finally solved in Byte Buddy and will be fixed with the next release!

**Comment by superyuuki** [comment]
I think i've accidentally revived this issue. I have

```
interface IndividualMotorComponent {
   int id();
} 
interface OtherMotorComponent {
  int someotherId();
}
interface MotorComponent extends OtherMotorComponent, IndividualMotorComponent {
   
}
```

I go to parse the methods on MotorComponent like so

```

for (Method method : useClass.getMethods()) {
  //add to a list
}
for (Method method : list) {
   builder = builder
                        .method(ElementMatchers.named(values.getKey().getName()))
                        .intercept(supplierInvoke);
}
```
where supplierInvoke is a MethodCall. This works for all interfaces that do not extend other interfaces (If i did this with IndividualMotorComponent it seemed to work okay), but when I use the MotorComponent interface i get the following:

```
Error at xyz.auriium.mattlib2.rev.HardwareREV.createSparkmax(HardwareREV.java:16): Unhandled exception: java.lang.AbstractMethodError: Receiver class xyz.auriium.mattlib2.hardware.config.MotorComponent$Delegated_e4d225ef does not define or inherit an implementation of the resolved method 'int id()' of interface xyz.auriium.mattlib2.hardware.config.MotorComponent.
```

Testing reveals that the method.intercept code is being called for the method id(), but when printed the Method displays as so:

```
public abstract int xyz.auriium.mattlib2.hardware.config.IndividualMotorComponent.id() implemented with 0
```
I do not know why this occurs but maybe bytebuddy is looking for a Method id specifically defined on MotorComponent, but isn't recognizing that IndividualMotorComponent defines id and so therefore MotorComponent should inherit that implementation?

**Comment by lmartelli** [comment]
I believe I have the same issue with with version 1.12.19.
I have several interfaces that extends a base interface, overriding a method with default method like this :
```
public interface HoteRepository extends JpaRepository<Hote, Integer> {
	@NotNull
	default Hote getById(int id) {
		return RepoUtils.requireEntity(this::findById, id, Hote.class);
	}

	@Query("select e from Hote e where e.id = :id")
	Optional<Hote> findById(@Param("id") Integer id);
}
```
When I stub the `getById()` method with Mockito, it sometime does not work (stubbed method returns null), and the cause appears to be that the `@Origin invokedMethod` is `JpaRepository.getById()` instead of `HoteRepository.getById()`.

**Comment by raphw** [comment]
I cannot reproduce this. Do you have an example project? I get the following to run correctly:

    interface Base {
        int foo();
    }

    interface Extension extends Base {
        @Override
        int foo();
    }

    public static void main(String[] args) {
        MethodGraph.Linked mg = MethodGraph.Compiler.DEFAULT.compile(TypeDescription.ForLoadedType.of(Extension.class));
        System.out.println(mg.listNodes().get(0).getRepresentative());
    }
